src/main/scala/mesosphere/marathon/core/storage/StorageConfig.scala

```diff
@@ -0,0 +1,316 @@
+    password: Option[String],
+    retries: Int,
+    enableCompression: Boolean,
+    compressionThreshold: ConfigMemorySize) extends LegacyStorageConfig {
+
+  private val sessionTimeoutTw = {
+    com.twitter.util.Duration(sessionTimeout.toMillis, TimeUnit.MILLISECONDS)
+  }
+  protected def store: PersistentStore = {
+    import com.twitter.util.JavaTimer
+    import com.twitter.zk.{ AuthInfo, NativeConnector, ZkClient }
+
+    val authInfo = (username, password) match {
+      case (Some(user), Some(pass)) => Some(AuthInfo.digest(user, pass))
+      case _ => None
+    }
+
+    val creationAcl = (username, password) match {
+      case (Some(user), Some(pass)) =>
+        ZooDefs.Ids.CREATOR_ALL_ACL
+      case _ =>
+        ZooDefs.Ids.OPEN_ACL_UNSAFE
+    }
+
+    val connector = NativeConnector(zkHosts, None, sessionTimeoutTw, new JavaTimer(isDaemon = true), authInfo)
+
+    val client = ZkClient(connector)
+      .withAcl(creationAcl)
+      .withRetries(retries)
+    val compressionConf = CompressionConf(enableCompression, compressionThreshold.toBytes)
+    new ZKStore(client, client(zkPath), compressionConf)
+  }
+}
+
+object ClassicZk {
+  def apply(config: AllConf): ClassicZk =
+    ClassicZk(
+      maxVersions = config.zooKeeperMaxVersions.get.get,
+      enableCache = config.storeCache.get.get,
+      sessionTimeout = config.zkSessionTimeoutDuration,
+      zkHosts = config.zkHosts,
+      zkPath = config.zkPath,
+      username = config.zkUsername,
+      password = config.zkPassword,
+      retries = 3,
+      enableCompression = config.zooKeeperCompressionEnabled.get.get,
+      compressionThreshold = ConfigMemorySize.ofBytes(config.zooKeeperCompressionThreshold.get.get))
+
+  def apply(config: Config): ClassicZk = {
+    // scalastyle:off
+    ClassicZk(
+      maxVersions = config.int("max-versions", 25),
+      enableCache = config.bool("enable-cache", true),
+      sessionTimeout = config.duration("session-timeout", 10.seconds),
+      zkHosts = {
+        val hosts = config.stringList("hosts")
+        if (hosts.nonEmpty) hosts.mkString(",")
+        else "localhost:2181"
+      },
+      zkPath = config.string("path", "marathon"),
+      username = config.optionalString("username"),
+      password = config.optionalString("password"),
+      retries = config.int("retries", 3),
+      enableCompression = config.bool("enable-compression", true),
+      compressionThreshold = config.memorySize("compression-threshold", ConfigMemorySize.ofBytes(64 * 1024))
+    )
+    // scalastyle:on
+  }
+}
+
+case class MesosZk(
+  maxVersions: Int,
+    enableCache: Boolean,
+    zkHosts: String,
+    zkPath: String,
+    timeout: Duration) extends LegacyStorageConfig {
+  def store: PersistentStore = {
+    val state = new ZooKeeperState(
+      zkHosts,
+      timeout.toMillis,
+      TimeUnit.MILLISECONDS,
+      zkPath
+    )
+    new MesosStateStore(state, timeout)
+  }
+}
+
+object MesosZk {
+  def apply(config: AllConf): MesosZk =
```

Comment:
Should take the specific Config?
Idea: AllConf should never be used in the components.